Add support for private incidents

cimi opened the pull request:
Allows marking incidents to signal they contain private information.
This can be done on creation or later when editing the incident.
Once the flag is set to true it cannot be undone.

Private incidents are not omitted when listing all incidents.
The incident summary, impact, severity etc. can still be edited from
 Slack after an incident is made private.

Incident details can no longer be retrieved from the API once an
incident is made private (always returns HTTP 403).

Marking an incident as private does not automatically change the comms
channel privacy settings, this needs to be done manually.

<img width="670" alt="Screenshot 2019-11-04 at 11 52 09" src="https://user-images.githubusercontent.com/348973/68122715-e21bc900-ff02-11e9-8fc0-916afbc2b2ad.png">

mattrco commented on response/slack/action_handlers.py:
```diff
@@ -53,41 +53,53 @@ def handle_create_comms_channel(ac: ActionContext):
+    if not ac.incident.private:
+        dialog_elements.append(
+            SelectWithOptions(
+                [("Yes (CANNOT BE UNDONE)", "yes")],
```
Very minor tweak but I'd put `Yes (can't be undone!)` for slightly friendlier text.

mattrco commented on response/core/serializers.py:
```diff
@@ -124,5 +125,8 @@ def update(self, instance, validated_data):
         instance.summary = validated_data.get("summary", instance.summary)
         instance.severity = validated_data.get("severity", instance.severity)
 
+        if not instance.private:
```
This makes sense, I'd probably add a small comment on what this does (only allows setting private to true).

mattrco commented:
🐛 There's a stray `__init__.pyc` here.

mattrco commented on tests/test.py:
```diff
@@ -220,3 +221,70 @@ def test_edit_incident(post_from_slack_api, mock_slack):
+    )
+
+    newsummary = "Testing editing incidents - after"
+    data = {
```
May be better to explicitly set private=False in this payload as that's being tested implicitly.

mattrco commented on response/slack/views.py:
```diff
@@ -69,6 +69,16 @@ def slash_command(request):
 
+    dialog.add_element(
+        SelectWithOptions(
+            [("Yes (CANNOT BE UNDONE)", "yes")],
```
Same minor comment re. wording

mattrco commented on tests/test.py:
```diff
@@ -220,3 +221,70 @@ def test_edit_incident(post_from_slack_api, mock_slack):
+
+    newsummary = "Testing editing incidents - after"
+    data = {
+        "payload": json.dumps(
```
Is it worth explicitly trying to set private=False in this payload?

mattrco commented:
@cimi this is great, thank you!

Although we've locked down the `incident_doc` view, private incidents will still be visible through the API (which is also what our internal UI uses to render data). We'll need to add a filter (`private=False`) to some of the view functions in `core/views.py` to achieve this. Happy to help with this if you'd like 👍

cimi commented:
Thanks for the review @mattrco! I've addressed your comments 😄

> We'll need to add a filter (private=False) to some of the view functions in `core/views.py` to achieve this. Happy to help with this if you'd like 👍

Should we do this in a separate PR? I'm at a conference today and tomorrow, if you have time to do the API update it would be great - we can release this feature sooner. No worries if not, I'll do it early next week.

mattrco commented:
@cimi I'm happy to merge this as-is and we can tackle the API later - although it's not quite complete, I think we need to do some test refactoring at the same time as it isn't very clear how the API views are tested vs. the HTML response.
